daemon/cdi_linux.go

```diff
@@ -0,0 +1,63 @@
+package daemon
+
+import (
+	"fmt"
+
+	"github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	"github.com/docker/docker/pkg/capabilities"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/sirupsen/logrus"
+)
+
+// RegisterCDIDriver registers the CDI device driver.
+// The driver injects CDI devices into an incoming OCI spec and is called for DeviceRequests associated with CDI devices.
+func RegisterCDIDriver(opts ...CDIOption) {
+	c := &cdiHandler{}
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	if len(c.specDirs) == 0 {
+		c.specDirs = cdi.DefaultSpecDirs
+	}
+
+	driver := &deviceDriver{
+		capset:     capabilities.Set{"cdi": struct{}{}},
+		updateSpec: c.injectCDIDevices,
+	}
+
+	registerDeviceDriver("cdi", driver)
+}
+
+// injectCDIDevices injects a set of CDI devices into the specified OCI specification.
+// Note that although the CDI specification can be applied to devices on platforms other than linux, the package used
+// currently only supports CDI device injection. If this restriction were to be relaxed, the function below should be usable as-is.
+func (c *cdiHandler) injectCDIDevices(s *specs.Spec, dev *deviceInstance) error {
+	if dev == nil {
+		return nil
+	}
+
+	cdiDeviceNames := dev.req.DeviceIDs
+	if len(cdiDeviceNames) == 0 {
+		return nil
+	}
+
+	registry := cdi.GetRegistry(
+		cdi.WithSpecDirs(c.specDirs...),
+	)
+	if err := registry.Refresh(); err != nil {
+		// We don't consider registry refresh failure a fatal error.
+		// For instance, a dynamically generated invalid CDI Spec file for
+		// any particular vendor shouldn't prevent injection of devices of
+		// different vendors. CDI itself knows better and it will fail the
+		// injection if necessary.
+		logrus.WithError(err).Debug("Refreshing the CDI registry failed")
+	}
```

Comment:
This block is redundant, `cdi.GetRegistry()` already does a `Refresh` call when given options. And when not given options, there's nothing for Refresh to do anyway.

Reply:
The call to `Refresh()` is a quick way to get access to generated errors instead of having to manually construct a multi-error. I have reworked things a bit to not call `Refresh()` and to directly instantiate a `NewCache` instead.

Reply:
I have udpated the code to call `cache.GetErrors()` and to construct a multi-error from the result. This is only raised IF the device injection fails.